store/tikv/gcworker/gc_worker.go

```diff
@@ -1066,20 +1073,31 @@ func (w *GCWorker) resolveLocksPhysical(ctx context.Context, safePoint uint64) e
 			return errors.Trace(err)
 		}
 
-		// Remove clean stores from the set
-		for resolvedStore := range resolvedStores {
-			// Only stores that are both resolved and checked is clean.
-			// For each clean store, remove it from the stores set.
-			if _, ok := checkedStores[resolvedStore]; ok {
-				delete(stores, resolvedStore)
+		for store := range stores {
+			if _, ok := checkedStores[store]; ok {
+				// The store is resolved and checked.
+				if _, ok := resolvedStores[store]; ok {
+					delete(stores, store)
+				}
+				// The store is checked and has been resolved before.
+				if _, ok := dirtyStores[store]; !ok {
+					delete(stores, store)
+				}
 			}
 		}
+		dirtyStores = stores
 
 		// If there are still dirty stores, continue the loop to clean them again.
 		// Only dirty stores will be scanned in the next loop.
-		if len(stores) == 0 {
+		if len(dirtyStores) == 0 {
 			break
 		}
+		// Clear all collected locks of dirty stores to prevent it from being dirty again.
+		w.removeLockObservers(ctx, safePoint, dirtyStores)
```

Comment:
I found we can't do it. If the store recevies duplicated `removeLockObserver` and `registerLockObserver` between L1061 and L1071. The store is cleaned but the lock observer drops some locks.

Reply:
Maybe we can't retry here.

Reply:
Maybe we shouldn't do "remove and re-register" on stores which fail to be checked. Instead, can we just register for those newly added stores?

Reply:
Yes. If dirty stores are newly added, we can retry. If a store is dirty due to too many collected locks, we don't retry and falls back to legacy mode.